// @coze/chat-sdk/dist/lib-source/src/libs/ui-kit/atomic/file-chooser/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import cls from 'classnames';
import { chooseImage } from '@tarojs/taro';
import { Text } from '@tarojs/components';
import { useI18n } from "../../../provider";
import { SvgImage, SvgCamera } from '../svg';
import { Spacing } from '../spacing';
import styles from './index.module.less';
const Chooser = ({ svg, text, onClick }) => (_jsxs(Spacing, Object.assign({ vertical: true, className: styles.chooser, gap: 8, onClick: onClick }, { children: [svg, _jsx(Text, Object.assign({ className: styles.text }, { children: text }))] })));
export const FileChooser = ({ className, onChooseFile }) => {
    const i18n = useI18n();
    const onClick = (sourceType) => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        try {
            const res = yield chooseImage({
                count: 1,
                sizeType: ['original', 'compressed'],
                sourceType,
            });
            if (res === null || res === void 0 ? void 0 : res.errMsg) {
                throw {
                    errMsg: res.errMsg,
                };
            }
            onChooseFile === null || onChooseFile === void 0 ? void 0 : onChooseFile(Object.assign({}, res));
        }
        catch (error) {
            const err = error;
            if (err.errNo !== 10502 && !((_a = err.errMsg) === null || _a === void 0 ? void 0 : _a.includes('cancel'))) {
                /* showToast({
                  content: i18n.t("chooseImageError"),
                  icon: "error",
                  duration: 3000,
                });*/
            }
        }
    });
    return (_jsxs(Spacing, Object.assign({ className: cls(styles.container, className), gap: 8 }, { children: [_jsx(Chooser, { svg: _jsx(SvgImage, { width: 24, height: 24 }), text: i18n.t('chooseImage'), onClick: () => {
                    onClick(['album']);
                } }), _jsx(Chooser, { svg: _jsx(SvgCamera, { width: 24, height: 24 }), text: i18n.t('chooseCamera'), onClick: () => {
                    onClick(['camera']);
                } })] })));
};
