// @coze/chat-sdk/dist/lib-source/src/libs/ui-kit/atomic/md-stream/render/block/code/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Fragment, useMemo } from 'react';
import cls from 'classnames';
import { showToast } from '@tarojs/taro';
import { View } from '@tarojs/components';
import { setClipboardData } from "../../../../../../utils";
import { Text } from '../../phrase/text';
import { Indicator } from '../../phrase/indicator';
import { useMdStreamEnableCodeBy4Space, useMdStreamGetMarkdown, useMdStreamI18n, } from '../../../context';
import styles from './index.module.less';
export const Code = ({ node }) => {
    var _a, _b;
    const i18n = useMdStreamI18n();
    const children = ((_a = node === null || node === void 0 ? void 0 : node.children) === null || _a === void 0 ? void 0 : _a.length)
        ? node === null || node === void 0 ? void 0 : node.children
        : [
            {
                type: 'text',
                value: node.value,
            },
        ];
    const enableCodeBy4Space = useMdStreamEnableCodeBy4Space();
    const markdown = useMdStreamGetMarkdown();
    const isShowCode = useMemo(() => {
        var _a, _b, _c;
        if (enableCodeBy4Space) {
            return true;
        }
        const offset = ((_a = node.position) === null || _a === void 0 ? void 0 : _a.start.offset) || 0;
        if (!((_b = node.position) === null || _b === void 0 ? void 0 : _b.start.offset) && ((_c = node.position) === null || _c === void 0 ? void 0 : _c.start.offset) !== 0) {
            return true;
        }
        else if (markdown.substring(offset, offset + 4) === '    ') {
            return false;
        }
        return true;
    }, [enableCodeBy4Space, markdown, (_b = node.position) === null || _b === void 0 ? void 0 : _b.start.offset]);
    return (_jsxs(View, Object.assign({ className: cls(styles['code-container'], {
            [styles.code]: isShowCode,
        }) }, { children: [_jsxs(View, Object.assign({ className: styles.header }, { children: [_jsx(View, Object.assign({ className: styles.lang }, { children: node.lang })), _jsx(View, Object.assign({ className: styles.copy, onClick: () => {
                            setClipboardData({
                                data: node.value,
                                fail() {
                                    showToast({
                                        title: (i18n === null || i18n === void 0 ? void 0 : i18n.t('copyFailed')) || '',
                                        icon: 'error',
                                    });
                                },
                                success(isUseWeb) {
                                    if (isUseWeb) {
                                        showToast({
                                            title: (i18n === null || i18n === void 0 ? void 0 : i18n.t('copySuccess')) || '',
                                            icon: 'success',
                                        });
                                    }
                                },
                            });
                        } }, { children: i18n === null || i18n === void 0 ? void 0 : i18n.t('copyCode') }))] })), _jsx(View, Object.assign({ className: styles.content }, { children: children === null || children === void 0 ? void 0 : children.map((item, index) => (_jsxs(Fragment, { children: [item.type === 'text' && (_jsx(Text, { node: item, className: styles.text })), item.type === 'indicator' && _jsx(Indicator, { node: item })] }, index))) }))] })));
};
