// @coze/chat-sdk/dist/lib-source/src/chatflow/helper/chat-service.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { getCdnUrl, logger, MiniChatError, MiniChatErrorCode, } from "../../libs/utils";
import { ChatService, } from "../../libs";
import { getConnectorId } from './get-connector-id';
// eslint-disable-next-line complexity
export const getCustomAppInfo = (chatFlowProps) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    let bgInfo;
    if (((_a = chatFlowProps.areaUi) === null || _a === void 0 ? void 0 : _a.bgInfo) && ((_c = (_b = chatFlowProps.areaUi) === null || _b === void 0 ? void 0 : _b.bgInfo) === null || _c === void 0 ? void 0 : _c.imgUrl)) {
        bgInfo = {
            imgUrl: chatFlowProps.areaUi.bgInfo.imgUrl,
            themeColor: chatFlowProps.areaUi.bgInfo.themeColor,
            gradientPosition: {
                left: 0,
                right: 0,
            },
            canvasPosition: {
                width: 0,
                height: 0,
                left: 0,
                top: 0,
            },
        };
    }
    return {
        name: (_d = chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project) === null || _d === void 0 ? void 0 : _d.name,
        icon_url: (_e = chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project) === null || _e === void 0 ? void 0 : _e.iconUrl,
        onboarding_info: {
            prologue: ((_g = (_f = chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project) === null || _f === void 0 ? void 0 : _f.onBoarding) === null || _g === void 0 ? void 0 : _g.prologue) || '',
            suggested_questions: ((_j = (_h = chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project) === null || _h === void 0 ? void 0 : _h.onBoarding) === null || _j === void 0 ? void 0 : _j.suggestions) || [],
        },
        description: (_k = chatFlowProps.project) === null || _k === void 0 ? void 0 : _k.desc,
        suggestPromoteInfo: (_l = chatFlowProps.project) === null || _l === void 0 ? void 0 : _l.suggestPromoteInfo,
        bgInfo: bgInfo
            ? {
                pc: bgInfo,
                mobile: bgInfo,
            }
            : undefined,
    };
};
// eslint-disable-next-line complexity
export const getBgInfo = (resp) => {
    var _a, _b, _c, _d, _e, _f;
    if (resp === null || resp === void 0 ? void 0 : resp.image_url) {
        return {
            canvasPosition: {
                height: ((_a = resp.canvas_position) === null || _a === void 0 ? void 0 : _a.height) || 0,
                left: ((_b = resp.canvas_position) === null || _b === void 0 ? void 0 : _b.left) || 0,
                top: ((_c = resp.canvas_position) === null || _c === void 0 ? void 0 : _c.top) || 0,
                width: ((_d = resp.canvas_position) === null || _d === void 0 ? void 0 : _d.width) || 0,
            },
            gradientPosition: {
                left: ((_e = resp.gradient_position) === null || _e === void 0 ? void 0 : _e.left) || 0,
                right: ((_f = resp.gradient_position) === null || _f === void 0 ? void 0 : _f.right) || 0,
            },
            imgUrl: resp.image_url || resp.origin_image_url || '',
            themeColor: resp.theme_color || '',
        };
    }
};
// eslint-disable-next-line complexity
export const combineRequestOptions = (appData, workflowData, chatFlowProps) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    const { icon_url: iconUrl, name = '' } = appData || {};
    const appInfoResult = {
        name,
        icon_url: iconUrl,
        description: '',
        create_time: 0,
        update_time: 0,
        version: '',
    };
    appInfoResult.onboarding_info = { prologue: '', suggested_questions: [] };
    appInfoResult.name =
        (workflowData === null || workflowData === void 0 ? void 0 : workflowData.name) ||
            appInfoResult.name ||
            ((_a = chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project) === null || _a === void 0 ? void 0 : _a.defaultName);
    appInfoResult.icon_url =
        ((_b = workflowData === null || workflowData === void 0 ? void 0 : workflowData.avatar) === null || _b === void 0 ? void 0 : _b.image_url) ||
            appInfoResult.icon_url ||
            (chatFlowProps === null || chatFlowProps === void 0 ? void 0 : chatFlowProps.project.defaultIconUrl);
    appInfoResult.onboarding_info = {
        prologue: ((_c = workflowData === null || workflowData === void 0 ? void 0 : workflowData.onboarding_info) === null || _c === void 0 ? void 0 : _c.prologue) || '',
        display_all_suggestions: (_d = workflowData === null || workflowData === void 0 ? void 0 : workflowData.onboarding_info) === null || _d === void 0 ? void 0 : _d.display_all_suggestions,
        suggested_questions: ((_f = (_e = workflowData === null || workflowData === void 0 ? void 0 : workflowData.onboarding_info) === null || _e === void 0 ? void 0 : _e.suggested_questions) === null || _f === void 0 ? void 0 : _f.filter(item => !!item)) || [],
    };
    appInfoResult.suggestPromoteInfo = (workflowData === null || workflowData === void 0 ? void 0 : workflowData.suggest_reply_info)
        ? {
            suggestReplyMode: (_g = workflowData === null || workflowData === void 0 ? void 0 : workflowData.suggest_reply_info) === null || _g === void 0 ? void 0 : _g.suggest_reply_mode,
            customizedSuggestPrompt: (_h = workflowData === null || workflowData === void 0 ? void 0 : workflowData.suggest_reply_info) === null || _h === void 0 ? void 0 : _h.customized_suggest_prompt,
        }
        : undefined;
    appInfoResult.bgInfo = {
        pc: getBgInfo((_j = workflowData === null || workflowData === void 0 ? void 0 : workflowData.background_image_info) === null || _j === void 0 ? void 0 : _j.web_background_image),
        mobile: getBgInfo((_k = workflowData === null || workflowData === void 0 ? void 0 : workflowData.background_image_info) === null || _k === void 0 ? void 0 : _k.mobile_background_image),
    };
    appInfoResult.voiceInfo = {
        isTextToVoiceEnable: (_l = workflowData === null || workflowData === void 0 ? void 0 : workflowData.audio_config) === null || _l === void 0 ? void 0 : _l.is_text_to_voice_enable,
        voiceConfigMap: (_m = workflowData === null || workflowData === void 0 ? void 0 : workflowData.audio_config) === null || _m === void 0 ? void 0 : _m.voice_config_map,
    };
    appInfoResult.inputMode = {
        default: (_o = workflowData === null || workflowData === void 0 ? void 0 : workflowData.user_input_config) === null || _o === void 0 ? void 0 : _o.default_input_mode,
    };
    return appInfoResult;
};
export class ChatFlowService extends ChatService {
    constructor(props, chatFlowProps) {
        super(props);
        this.chatFlowProps = chatFlowProps;
    }
    updateChatFlowProps(chatFlowProps) {
        this.chatFlowProps = chatFlowProps;
    }
    createNewConversation() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._createNewConversation(false);
        });
    }
    // eslint-disable-next-line complexity
    getAppInfo() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        return __awaiter(this, void 0, void 0, function* () {
            const connectorId = getConnectorId(this.chatFlowProps);
            const isWebSdk = ((_b = (_a = this.chatFlowProps) === null || _a === void 0 ? void 0 : _a.project) === null || _b === void 0 ? void 0 : _b.mode) === 'websdk';
            const [appRes, workflowRes] = yield Promise.allSettled([
                isWebSdk
                    ? (_c = this.apiClient) === null || _c === void 0 ? void 0 : _c.get(`/v1/apps/${(_e = (_d = this.chatFlowProps) === null || _d === void 0 ? void 0 : _d.project) === null || _e === void 0 ? void 0 : _e.id}?connector_id=${connectorId}`)
                    : null,
                ((_g = (_f = this.chatFlowProps) === null || _f === void 0 ? void 0 : _f.workflow) === null || _g === void 0 ? void 0 : _g.id)
                    ? (_h = this.apiClient) === null || _h === void 0 ? void 0 : _h.get(`/v1/workflows/${(_k = (_j = this.chatFlowProps) === null || _j === void 0 ? void 0 : _j.workflow) === null || _k === void 0 ? void 0 : _k.id}?${[
                        `connector_id=${connectorId}`,
                        `is_debug=${((_m = (_l = this.chatFlowProps) === null || _l === void 0 ? void 0 : _l.project) === null || _m === void 0 ? void 0 : _m.mode) === 'draft' ? 'true' : ''}`,
                        `caller=${((_p = (_o = this.chatFlowProps) === null || _o === void 0 ? void 0 : _o.project) === null || _p === void 0 ? void 0 : _p.caller) || ''}`,
                    ].join('&')}`)
                    : null,
            ]);
            const appData = (appRes === null || appRes === void 0 ? void 0 : appRes.status) === 'fulfilled' ? (_q = appRes === null || appRes === void 0 ? void 0 : appRes.value) === null || _q === void 0 ? void 0 : _q.data : undefined;
            const workflowData = (workflowRes === null || workflowRes === void 0 ? void 0 : workflowRes.status) === 'fulfilled'
                ? (_s = (_r = workflowRes === null || workflowRes === void 0 ? void 0 : workflowRes.value) === null || _r === void 0 ? void 0 : _r.data) === null || _s === void 0 ? void 0 : _s.role
                : undefined;
            logger.debug('Get Workflow Info: ', { appRes, workflowRes });
            if (isWebSdk && !appData) {
                throw new MiniChatError(MiniChatErrorCode.SDK_API_APP_UnPublished, 'The app is not published');
            }
            // 默认数据不用合并进去
            const appInfo = combineRequestOptions(appData, workflowData);
            if (((_u = (_t = this.chatFlowProps) === null || _t === void 0 ? void 0 : _t.workflow) === null || _u === void 0 ? void 0 : _u.id) && !appInfo.icon_url) {
                appInfo.icon_url = getCdnUrl((_v = this.chatFlowProps.setting) === null || _v === void 0 ? void 0 : _v.cdnBaseUrlPath, 'assets/imgs/chatflow-logo.png');
            }
            logger.debug('Get Workflow Info(Combine):', appInfo);
            return Object.assign({ appId: this.appId, type: this.chatType }, appInfo);
        });
    }
    getOrCreateConversationId() {
        return this._createNewConversation(true);
    }
    _createNewConversation(isCreateNew = false) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        return __awaiter(this, void 0, void 0, function* () {
            if (((_b = (_a = this.chatFlowProps) === null || _a === void 0 ? void 0 : _a.project) === null || _b === void 0 ? void 0 : _b.type) === 'bot') {
                const { id: conversationId, last_section_id: sectionId = '' } = yield this.apiClient.conversations.create({
                    // @ts-expect-error -- linter-disable-autofix
                    connector_id: this.connectorId,
                });
                return { conversationId, sectionId };
            }
            else {
                const { id: conversationId, last_section_id: sectionId = '' } = yield this.apiClient.conversations.create({
                    app_id: this.appId,
                    conversation_name: (_d = (_c = this.chatFlowProps) === null || _c === void 0 ? void 0 : _c.project) === null || _d === void 0 ? void 0 : _d.conversationName,
                    get_or_create: isCreateNew,
                    workflow_id: (_f = (_e = this.chatFlowProps) === null || _e === void 0 ? void 0 : _e.workflow) === null || _f === void 0 ? void 0 : _f.id,
                    draft_mode: ((_h = (_g = this.chatFlowProps) === null || _g === void 0 ? void 0 : _g.project) === null || _h === void 0 ? void 0 : _h.mode) === 'draft',
                    connector_id: getConnectorId(this.chatFlowProps),
                });
                return { conversationId, sectionId };
            }
        });
    }
    // eslint-disable-next-line complexity
    asyncChat(params, options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        const bodyData = {};
        bodyData.additional_messages = params.additional_messages || [];
        bodyData.connector_id = params.connector_id;
        bodyData.conversation_id = params.conversation_id;
        bodyData.workflow_id = (_b = (_a = this.chatFlowProps) === null || _a === void 0 ? void 0 : _a.workflow) === null || _b === void 0 ? void 0 : _b.id;
        bodyData.parameters = (_d = (_c = this.chatFlowProps) === null || _c === void 0 ? void 0 : _c.workflow) === null || _d === void 0 ? void 0 : _d.parameters;
        bodyData.execute_mode =
            ((_f = (_e = this.chatFlowProps) === null || _e === void 0 ? void 0 : _e.project) === null || _f === void 0 ? void 0 : _f.mode) === 'draft' ? 'DEBUG' : undefined;
        bodyData.app_id =
            ((_h = (_g = this.chatFlowProps) === null || _g === void 0 ? void 0 : _g.project) === null || _h === void 0 ? void 0 : _h.type) === 'app'
                ? (_k = (_j = this.chatFlowProps) === null || _j === void 0 ? void 0 : _j.project) === null || _k === void 0 ? void 0 : _k.id
                : undefined;
        bodyData.bot_id =
            ((_m = (_l = this.chatFlowProps) === null || _l === void 0 ? void 0 : _l.project) === null || _m === void 0 ? void 0 : _m.type) === 'bot'
                ? (_p = (_o = this.chatFlowProps) === null || _o === void 0 ? void 0 : _o.project) === null || _p === void 0 ? void 0 : _p.id
                : undefined;
        bodyData.connector_id = getConnectorId(this.chatFlowProps);
        bodyData.ext = {
            _caller: (_r = (_q = this.chatFlowProps) === null || _q === void 0 ? void 0 : _q.project) === null || _r === void 0 ? void 0 : _r.caller,
            user_id: params.user_id,
            enable_card: 'true',
        };
        bodyData.suggest_reply_info = params.suggestPromoteInfo
            ? {
                suggest_reply_mode: (_s = params.suggestPromoteInfo) === null || _s === void 0 ? void 0 : _s.suggestReplyMode,
                customized_suggest_prompt: (_t = params.suggestPromoteInfo) === null || _t === void 0 ? void 0 : _t.customizedSuggestPrompt,
            }
            : undefined;
        const optionNew = options || {};
        optionNew.headers = Object.assign(Object.assign({}, ((options === null || options === void 0 ? void 0 : options.headers) || {})), (((_v = (_u = this.chatFlowProps) === null || _u === void 0 ? void 0 : _u.workflow) === null || _v === void 0 ? void 0 : _v.header) || {}));
        return this.apiClient.workflows.chat.stream(bodyData, optionNew);
    }
}
