// @coze/chat-sdk/dist/lib-source/src/chatflow/helper/get-token.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { request } from '@tarojs/taro';
import { logger } from "../../libs/utils";
export const getToken = (props) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c;
    let token = (_a = props.auth) === null || _a === void 0 ? void 0 : _a.token;
    let refreshToken = (_b = props.auth) === null || _b === void 0 ? void 0 : _b.refreshToken;
    if (((_c = props.auth) === null || _c === void 0 ? void 0 : _c.type) === 'internal') {
        token = (yield getTokenInternal()) || '';
        refreshToken = getTokenInternal;
    }
    return { token, refreshToken };
});
function getTokenInternal() {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const result = yield request({
                url: '/api/permission_api/coze_web_app/impersonate_coze_user',
                method: 'POST',
                dataType: 'json',
                timeout: 30000,
                data: {},
            });
            return (_b = (_a = result.data) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.access_token;
        }
        catch (_err) {
            logger.error('getTokenInternal error', _err);
            return '';
        }
    });
}
