// @coze/chat-sdk/dist/lib-source/src/libs/utils/event-source/event-source.js
import { Events } from '@tarojs/taro';
import { MiniChatError } from "..";
export const EventJsonDataType = Symbol('eventJsonDataType');
export var MessageEvent;
(function (MessageEvent) {
    MessageEvent["ERROR"] = "error";
    MessageEvent["CLOSE"] = "close";
    MessageEvent["OPEN"] = "open";
    MessageEvent["MESSAGE"] = "message";
})(MessageEvent || (MessageEvent = {}));
export class EventSourceBase {
    constructor({ url, method = 'GET', header = {}, data }) {
        this.isClosed = false;
        this.event = new Events();
        this.url = url;
        this.method = method;
        this.header = header;
        this.data = data;
    }
    sendMessage() {
        this._sendMessage();
    }
    _sendMessage() {
        throw new MiniChatError(-1, 'not implemented');
    }
    close() {
        if (this.isClosed) {
            return false;
        }
        this.isClosed = true;
        this.event.trigger(MessageEvent.CLOSE);
        this.event.off();
        return true;
    }
    onError(onError) {
        this.event.on(MessageEvent.ERROR, err => {
            onError(new MiniChatError(-1, err.errMsg));
        });
    }
    onClose(onClose) {
        this.event.on(MessageEvent.CLOSE, onClose);
    }
    onMessage(onMessage) {
        this.event.on(MessageEvent.MESSAGE, onMessage);
    }
    onOpen(onOpen) {
        this.event.on(MessageEvent.OPEN, onOpen);
    }
}
