// @coze/chat-sdk/dist/lib-source/src/chatflow/hooks/use-init-chat.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useEffect, useMemo, useRef, useState } from 'react';
import { logger } from "../../libs/utils";
import { ChatType } from "../../libs/types";
import { useUpdateEffect } from "../../libs/hooks";
import { getToken } from '../helper/get-token';
import { getConnectorId } from '../helper/get-connector-id';
import { genErrorRender } from '../helper/gen-error-render';
import { ChatFlowService, getCustomAppInfo } from '../helper/chat-service';
import styles from './index.module.less';
export const useInitChat = (props) => {
    var _a, _b;
    const { project: projectInfo, userInfo } = props;
    const [hasReady, setHasReady] = useState(false);
    const [authProp, setAuthProp] = useState();
    const refService = useRef();
    // The parameters will be changed immediately, so we don't code it in useEffect.
    (_a = refService.current) === null || _a === void 0 ? void 0 : _a.updateChatFlowProps(props);
    const syncModifyChatProps = useMemo(
    // eslint-disable-next-line complexity
    () => {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
        return ({
            chat: Object.assign({ appId: projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.id, type: (projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.type) === 'app' ? ChatType.App : ChatType.Bot }, getCustomAppInfo(props)),
            user: userInfo,
            ui: {
                layout: ((_a = props.areaUi) === null || _a === void 0 ? void 0 : _a.layout) || 'pc',
                isMiniCustomHeader: (_b = props === null || props === void 0 ? void 0 : props.areaUi) === null || _b === void 0 ? void 0 : _b.isMiniCustomHeader,
                isReadonly: (_c = props === null || props === void 0 ? void 0 : props.areaUi) === null || _c === void 0 ? void 0 : _c.isDisabled,
                header: Object.assign({ isNeed: false }, (((_d = props === null || props === void 0 ? void 0 : props.areaUi) === null || _d === void 0 ? void 0 : _d.header) || {})),
                footer: Object.assign({ containerClassName: ((_e = props.project) === null || _e === void 0 ? void 0 : _e.mode) !== 'websdk'
                        ? styles['footer-container']
                        : undefined }, (_f = props === null || props === void 0 ? void 0 : props.areaUi) === null || _f === void 0 ? void 0 : _f.footer),
                loading: {
                    renderLoading: (_g = props === null || props === void 0 ? void 0 : props.areaUi) === null || _g === void 0 ? void 0 : _g.renderLoading,
                },
                error: {
                    renderError: genErrorRender(props),
                },
                chatSlot: {
                    base: {
                        maxWidth: 600,
                        isFrameAutoFocus: (_h = props === null || props === void 0 ? void 0 : props.areaUi) === null || _h === void 0 ? void 0 : _h.isFrameAutoFocus,
                    },
                    input: (_j = props === null || props === void 0 ? void 0 : props.areaUi) === null || _j === void 0 ? void 0 : _j.input,
                    clearContext: (_k = props === null || props === void 0 ? void 0 : props.areaUi) === null || _k === void 0 ? void 0 : _k.clearContext,
                    clearMessage: (_l = props === null || props === void 0 ? void 0 : props.areaUi) === null || _l === void 0 ? void 0 : _l.clearMessage,
                    uploadBtn: (_m = props === null || props === void 0 ? void 0 : props.areaUi) === null || _m === void 0 ? void 0 : _m.uploadBtn,
                    message: (_o = props === null || props === void 0 ? void 0 : props.areaUi) === null || _o === void 0 ? void 0 : _o.message,
                    messageWrapper: (_p = props === null || props === void 0 ? void 0 : props.areaUi) === null || _p === void 0 ? void 0 : _p.messageWrapper,
                },
            },
            eventCallbacks: {
                onImageClick: (_q = props === null || props === void 0 ? void 0 : props.eventCallbacks) === null || _q === void 0 ? void 0 : _q.onImageClick,
                onInitRetry: () => {
                    setHasReady(false);
                },
                onThemeChange: (_r = props === null || props === void 0 ? void 0 : props.eventCallbacks) === null || _r === void 0 ? void 0 : _r.onThemeChange,
                onLinkClick: (_s = props === null || props === void 0 ? void 0 : props.eventCallbacks) === null || _s === void 0 ? void 0 : _s.onLinkClick,
                message: (_t = props === null || props === void 0 ? void 0 : props.eventCallbacks) === null || _t === void 0 ? void 0 : _t.message,
            },
            setting: Object.assign(Object.assign({}, ((props === null || props === void 0 ? void 0 : props.setting) || {})), { onGetCustomChatService: chatServiceProps => {
                    refService.current = new ChatFlowService(chatServiceProps, props);
                    return refService.current;
                } }),
        });
    }, [props]);
    useEffect(() => {
        if (hasReady) {
            return;
        }
        // just check param for console, don't do any more;
        checkParam(props);
        (() => __awaiter(void 0, void 0, void 0, function* () {
            // getToken will never throw an exception, so we don't need to catch it;
            const { token, refreshToken } = yield getToken(props);
            setAuthProp({
                // don't need to check token, because it will be checked in chat framework
                token: token,
                onRefreshToken: refreshToken,
                connectorId: getConnectorId(props),
            });
            setHasReady(true);
        }))();
    }, [hasReady]);
    useUpdateEffect(() => {
        logger.info('useInitChat reRender');
        setHasReady(false);
    }, [
        projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.id,
        projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.type,
        projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.conversationName,
        projectInfo === null || projectInfo === void 0 ? void 0 : projectInfo.mode,
        (_b = props === null || props === void 0 ? void 0 : props.workflow) === null || _b === void 0 ? void 0 : _b.id,
    ]);
    return {
        hasReady,
        chatProps: Object.assign(syncModifyChatProps, { auth: authProp }),
    };
};
function checkParam(props) {
    var _a, _b, _c, _d;
    let error;
    if (((_a = props === null || props === void 0 ? void 0 : props.project) === null || _a === void 0 ? void 0 : _a.type) === 'bot') {
        if (((_b = props === null || props === void 0 ? void 0 : props.project) === null || _b === void 0 ? void 0 : _b.mode) !== 'draft') {
            logger.error('mode must be draft when project type is bot');
        }
    }
    else {
        if (((_c = props === null || props === void 0 ? void 0 : props.auth) === null || _c === void 0 ? void 0 : _c.type) !== 'internal') {
            if (!((_d = props === null || props === void 0 ? void 0 : props.auth) === null || _d === void 0 ? void 0 : _d.token)) {
                logger.error('token is required when auth type is not internal');
            }
        }
    }
    return error;
}
