// @coze/chat-sdk/dist/lib-source/src/libs/services/use-clear-context.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useCallback } from 'react';
import { logger, showToast } from "../utils";
import { useApiClientStore } from "../provider";
import { UIEventType } from '../types';
import { useConversationStore, useChatStatusStore, useI18n, useUiEventStore, } from '../provider/context/chat-store-context';
export const useClearContext = () => {
    const { conversationId, setSectionId } = useConversationStore(store => ({
        conversationId: store.id,
        setSectionId: store.setSectionId,
    }));
    const { getOpDisabledState, setIsClearingContext } = useChatStatusStore(store => ({
        setIsClearingContext: store.setIsClearingContext,
        getOpDisabledState: store.getOpDisabledState,
    }));
    const targetEventCenter = useUiEventStore(store => store.event);
    const chatEvent = useUiEventStore(store => store.event);
    const chatService = useApiClientStore(store => store.chatService);
    const i18n = useI18n();
    const clearContext = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        const { clearContext: disableState } = getOpDisabledState();
        if (!conversationId) {
            logger.error('clearContext, conversationId is empty');
            return;
        }
        if (disableState) {
            return;
        }
        setIsClearingContext(true);
        chatEvent.trigger(UIEventType.ChatSlotScrollToAnchorBottom);
        try {
            const { sectionId } = yield chatService.createNewSection(conversationId);
            setSectionId(sectionId);
        }
        catch (err) {
            showToast({
                content: i18n.t('clearContextFailed'),
                icon: 'error',
            }, targetEventCenter);
        }
        chatEvent.trigger(UIEventType.ChatSlotRemoveAnchorBottom);
        setIsClearingContext(false);
    }), [conversationId, targetEventCenter]);
    return {
        clearContext,
    };
};
