// @coze/chat-sdk/dist/lib-source/src/libs/services/use-create-prologue-message.js
import { useCallback, useMemo } from 'react';
import { RoleType } from '@coze/api';
import { useConversationStore, useChatInfoStore, useUserInfoStore, } from "../provider";
import { nanoid } from '../utils';
export const useCreatePrologueMessage = (nextSectionId) => {
    var _a, _b;
    const { chatInfo } = useChatInfoStore(store => ({
        chatInfo: store.info,
    }));
    const { conversationId } = useConversationStore(store => ({
        conversationId: store.id,
    }));
    const { userInfo } = useUserInfoStore(store => ({
        userInfo: store.info,
    }));
    const createAnswerMessage = useCallback((type, content) => ({
        id: '',
        conversation_id: conversationId || '',
        section_id: nextSectionId || '',
        bot_id: (chatInfo === null || chatInfo === void 0 ? void 0 : chatInfo.appId) || '',
        chat_id: '',
        role: RoleType.Assistant,
        type,
        localId: nanoid(),
        meta_data: {},
        content_type: 'text',
        created_at: 0,
        updated_at: 0,
        content,
    }), [conversationId, nextSectionId, chatInfo]);
    const localId = useMemo(() => nanoid(), []);
    const chatMessage = useMemo(() => {
        var _a, _b;
        return [
            createAnswerMessage('answer', (((_a = chatInfo === null || chatInfo === void 0 ? void 0 : chatInfo.onboarding_info) === null || _a === void 0 ? void 0 : _a.prologue) || '').replaceAll('{{user_name}}', (userInfo === null || userInfo === void 0 ? void 0 : userInfo.name) || '')),
            ...(((_b = chatInfo === null || chatInfo === void 0 ? void 0 : chatInfo.onboarding_info) === null || _b === void 0 ? void 0 : _b.suggested_questions) || []).map(item => createAnswerMessage('follow_up', item)),
        ];
    }, [chatInfo, conversationId, createAnswerMessage, userInfo]);
    const chatGroup = useMemo(() => ({
        id: localId,
        chatId: '',
        respMessages: chatMessage,
        sectionId: nextSectionId,
        isPrologue: true,
    }), [chatMessage, localId, nextSectionId]);
    return ((_a = chatInfo === null || chatInfo === void 0 ? void 0 : chatInfo.onboarding_info) === null || _a === void 0 ? void 0 : _a.prologue) &&
        ((_b = chatInfo === null || chatInfo === void 0 ? void 0 : chatInfo.onboarding_info) === null || _b === void 0 ? void 0 : _b.prologue) !== '\n' // chatflow里边 默认会有一个\n
        ? chatGroup
        : null;
};
