// @coze/chat-sdk/dist/lib-source/src/libs/services/use-send-message.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useCallback } from 'react';
import { showToast } from "../utils";
import { useApiClientStore, useUserInfoStore } from "../provider";
import { SendMessageEvent } from '../types';
import { useConversationStore, useChatStatusStore, useChatInfoStore, useI18n, useUiEventStore, useChatInputStore, useChatPropsStore, } from '../provider/context/chat-store-context';
import { getSendMessageHandler, RawMessageType, } from './helper/message';
import { usePersistCallback } from '../hooks';
// eslint-disable-next-line max-lines-per-function
export const useSendMessage = () => {
    const { setSendMessageService, conversationId, sectionId, popLastErrorChatGroup, } = useConversationStore(store => ({
        setSendMessageService: store.setSendMessageService,
        conversationId: store.id,
        sectionId: store.sectionId,
        popLastErrorChatGroup: store.popLastErrorChatGroup,
    }));
    const i18n = useI18n();
    const userInfo = useUserInfoStore(store => store.info);
    const botId = useChatInfoStore(store => store.id);
    const { getOpDisabledState, setIsDeleting } = useChatStatusStore(store => ({
        setIsDeleting: store.setIsDeleting,
        getOpDisabledState: store.getOpDisabledState,
    }));
    const chatInfo = useChatInfoStore(store => store.info);
    const setTaskList = useChatInputStore(store => store.setTaskList);
    const { connectorId, chatService } = useApiClientStore(store => ({
        connectorId: store.connectorId,
        chatService: store.chatService,
    }));
    const targetEventCenter = useUiEventStore(store => store.event);
    const onRequiresAction = useChatPropsStore(store => { var _a, _b; return (_b = (_a = store.eventCallbacks) === null || _a === void 0 ? void 0 : _a.message) === null || _b === void 0 ? void 0 : _b.onRequiresAction; });
    const checkCanSendMessage = useChatPropsStore(store => { var _a, _b, _c; return (_c = (_b = (_a = store.ui) === null || _a === void 0 ? void 0 : _a.chatSlot) === null || _b === void 0 ? void 0 : _b.input) === null || _c === void 0 ? void 0 : _c.checkCanSendMessage; });
    const sendMessage = usePersistCallback((rawMessage, historyMessages) => __awaiter(void 0, void 0, void 0, function* () {
        const { clearMessage: disableState } = getOpDisabledState();
        if (disableState) {
            return;
        }
        if (!botId || !conversationId) {
            return;
        }
        if (checkCanSendMessage) {
            if ((yield checkCanSendMessage(rawMessage)) === false) {
                return false;
            }
        }
        // Clear task message after message sended
        setTaskList({ taskList: [] });
        setIsDeleting(true);
        const sendMessageHandler = getSendMessageHandler({
            botId,
            chatService,
            conversationId,
            userId: (userInfo === null || userInfo === void 0 ? void 0 : userInfo.id) || '',
            connectorId,
            sectionId,
            chatInfo: chatInfo || undefined,
            i18n,
        });
        sendMessageHandler.on(SendMessageEvent.RequireAction, event => {
            onRequiresAction === null || onRequiresAction === void 0 ? void 0 : onRequiresAction({
                extra: {
                    requireAction: event.event,
                },
            });
        });
        sendMessageHandler.on(SendMessageEvent.Close, () => {
            setIsDeleting(false);
        });
        sendMessageHandler.on(SendMessageEvent.ReceiveComplete, event => {
            var _a;
            if (event.error) {
                showToast({
                    content: ((_a = event.error) === null || _a === void 0 ? void 0 : _a.getErrorMessageByI18n(i18n, {})) ||
                        i18n.t('sendFailed'),
                    icon: 'error',
                    duration: 2000,
                }, targetEventCenter);
            }
        });
        setSendMessageService(sendMessageHandler);
        sendMessageHandler.sendRawMessage(rawMessage, historyMessages);
    }));
    const sendTextMessage = useCallback((content) => __awaiter(void 0, void 0, void 0, function* () {
        return yield sendMessage({
            type: RawMessageType.TEXT,
            data: content,
        });
    }), [sendMessage]);
    const sendFileMessage = useCallback((files) => __awaiter(void 0, void 0, void 0, function* () {
        return yield sendMessage({
            type: RawMessageType.FILE,
            data: files,
        });
    }), [sendMessage]);
    const sendAudioMessage = useCallback((audio) => __awaiter(void 0, void 0, void 0, function* () {
        return yield sendMessage({
            type: RawMessageType.AUDIO,
            data: audio,
        });
    }), [sendMessage]);
    const reSendLastErrorMessage = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        var _a, _b;
        const chatMessageGroup = popLastErrorChatGroup();
        if ((_a = chatMessageGroup === null || chatMessageGroup === void 0 ? void 0 : chatMessageGroup.query) === null || _a === void 0 ? void 0 : _a.rawMessage) {
            return yield sendMessage((_b = chatMessageGroup === null || chatMessageGroup === void 0 ? void 0 : chatMessageGroup.query) === null || _b === void 0 ? void 0 : _b.rawMessage);
        }
    }), [sendMessage]);
    return {
        sendMessage,
        sendTextMessage,
        sendFileMessage,
        sendAudioMessage,
        reSendLastErrorMessage,
    };
};
