// @coze/chat-sdk/dist/lib-source/src/libs/services/use-chat-init.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useEffect, useState } from 'react';
import { convertToMinChatError, logger, MiniChatErrorCode, MiniChatError, } from "../utils";
import { useApiClientStore, useChatPropsContext, useChatInfoStore, } from "../provider";
import { useConversationStore } from '../provider/context/chat-store-context';
export const useChatInit = () => {
    const chatProps = useChatPropsContext();
    const { setChatError, setChatInfo, setChatIsLoading, error: chatError, } = useChatInfoStore(store => ({
        setChatError: store.setError,
        setChatInfo: store.setChatInfo,
        setChatIsLoading: store.setIsLoading,
        setCustomChatInfo: store.setCustomChatInfo,
        isLoadingChatInfo: store.isLoading,
        error: store.error,
    }));
    const setConversationDetail = useConversationStore(store => store.setConversationDetail);
    const chatService = useApiClientStore(store => store.chatService);
    const [retryTime, setRetryTime] = useState(0);
    useEffect(() => {
        if (!checkParams(chatProps)) {
            setChatError(new MiniChatError(-1, 'params error'));
            return;
        }
        let isAbort = false;
        setChatError(null);
        setChatIsLoading(true);
        (() => __awaiter(void 0, void 0, void 0, function* () {
            try {
                const [chatInfo, conversationDetail] = yield Promise.all([
                    chatService.getAppInfo(),
                    getInitConversationDetail(chatService),
                ]);
                if (!isAbort) {
                    setConversationDetail(conversationDetail, conversationDetail.messages);
                    setChatInfo(chatInfo);
                }
            }
            catch (err) {
                logger.error('get chat info error', err);
                const miniChatError = convertToMinChatError(err);
                setChatError(miniChatError);
            }
        }))();
        return () => {
            isAbort = true;
        };
    }, [retryTime, chatService]);
    return {
        retryChatInit: () => {
            if (chatError) {
                setRetryTime(retryTime + 1);
            }
        },
    };
};
/*
 * @param chatProps - The chat props.
 * @returns A boolean value.
 */
function checkParams(chatProps) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    if (!((_a = chatProps === null || chatProps === void 0 ? void 0 : chatProps.setting) === null || _a === void 0 ? void 0 : _a.apiBaseUrl)) {
        logger.error("Setting's apiBaseUrl must be provided");
        return false;
    }
    if (!((_b = chatProps === null || chatProps === void 0 ? void 0 : chatProps.setting) === null || _b === void 0 ? void 0 : _b.cdnBaseUrlPath)) {
        logger.warn("Setting's cdnBaseUrlPath is not provided; The chat will show a crash image");
    }
    if (!((_c = chatProps === null || chatProps === void 0 ? void 0 : chatProps.chat) === null || _c === void 0 ? void 0 : _c.appId)) {
        logger.error("Chat's appId must be provided");
        return false;
    }
    if (!((_d = chatProps === null || chatProps === void 0 ? void 0 : chatProps.auth) === null || _d === void 0 ? void 0 : _d.token)) {
        logger.error('Token  must be provided');
        return false;
    }
    if (!((_e = chatProps === null || chatProps === void 0 ? void 0 : chatProps.auth) === null || _e === void 0 ? void 0 : _e.onRefreshToken)) {
        logger.warn('onRefreshToken is not provided; The chat will not be able to refresh the token');
    }
    if (!((_f = chatProps === null || chatProps === void 0 ? void 0 : chatProps.user) === null || _f === void 0 ? void 0 : _f.id) ||
        !((_g = chatProps === null || chatProps === void 0 ? void 0 : chatProps.user) === null || _g === void 0 ? void 0 : _g.name) ||
        !((_h = chatProps === null || chatProps === void 0 ? void 0 : chatProps.user) === null || _h === void 0 ? void 0 : _h.avatar)) {
        logger.warn("User's property (id, name, avatar) is empty; The chat will show a default user");
        return true;
    }
    return true;
}
/*
 * Get the initial conversation detail
 * @param chatService - The chat service to use.
 * @returns The conversation detail(conversationId, sectionId, messages)
 */
function getInitConversationDetail(chatService) {
    return __awaiter(this, void 0, void 0, function* () {
        let conversationId = '';
        let sectionId = '';
        try {
            const res = yield chatService.getOrCreateConversationId();
            conversationId = res.conversationId;
            sectionId = res.sectionId;
        }
        catch (error) {
            logger.error('Get Message Error', error);
            const apiError = convertToMinChatError(error);
            throw new MiniChatError(apiError.code || MiniChatErrorCode.Custom_Conversation_Create_Error, apiError.msg || 'Conversation Create Error');
        }
        try {
            const messageListRes = yield chatService.getMessageList({
                conversationId,
            });
            return Object.assign({ id: conversationId, sectionId }, messageListRes);
        }
        catch (err) {
            throw new MiniChatError(MiniChatErrorCode.Custom_GetMessageList_Error, 'Get MessageList Error');
        }
    });
}
