// @coze/chat-sdk/dist/lib-source/src/libs/services/use-clear-message.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useCallback } from 'react';
import { showToast } from "../utils";
import { useApiClientStore } from "../provider";
import { useConversationStore, useChatStatusStore, useChatInfoStore, useI18n, useUiEventStore, } from '../provider/context/chat-store-context';
export const useClearMessage = () => {
    const { setNewConversationId, setSectionId } = useConversationStore(store => ({
        setNewConversationId: store.setNewConversationId,
        setSectionId: store.setSectionId,
    }));
    const i18n = useI18n();
    const appId = useChatInfoStore(store => store.id);
    const { getOpDisabledState, setIsDeleting } = useChatStatusStore(store => ({
        setIsDeleting: store.setIsDeleting,
        getOpDisabledState: store.getOpDisabledState,
    }));
    const targetEventCenter = useUiEventStore(store => store.event);
    const chatService = useApiClientStore(store => store.chatService);
    const clearMessage = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        const { clearMessage: disableState } = getOpDisabledState();
        if (disableState) {
            return;
        }
        setIsDeleting(true);
        try {
            const { conversationId, sectionId } = yield chatService.createNewConversation();
            setNewConversationId(conversationId);
            setSectionId(sectionId);
        }
        catch (err) {
            showToast({
                content: i18n.t('clearMessageFailed'),
                icon: 'error',
            }, targetEventCenter);
        }
        setIsDeleting(false);
    }), [appId, targetEventCenter]);
    return {
        clearMessage,
    };
};
