// @coze/chat-sdk/dist/lib-source/src/libs/framework/component/chat-scroll-view/index.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useMemo, useRef } from 'react';
import cls from 'classnames';
import { logger } from "../../../utils";
import { ScrollView, Spinning, ErrorRetry, ErrorRetryBtn, } from "../../../ui-kit";
import { UIEventType } from "../../../types";
import { useScrollMore } from "../../../services";
import { useUiEventStore } from "../../../provider/context/chat-store-context";
import { useConversationStore, useI18n } from "../../../provider";
import { useUpdateEffect } from "../../../hooks";
import styles from './index.module.less';
const ChatTopStatus = ({ onScrollToUpper }) => {
    const i18n = useI18n();
    const { prevError, prevHasMore } = useConversationStore(store => ({
        prevError: store.prevError,
        prevHasMore: store.prevHasMore,
    }));
    const isNeedPrevLoadMore = prevHasMore;
    return (_jsx(_Fragment, { children: prevError ? (_jsx(ErrorRetry, { errorText: i18n.t('messageListRetry', {
                retry: (_jsx(ErrorRetryBtn, { retryText: i18n.t('retryBtn'), onClick: () => {
                        onScrollToUpper === null || onScrollToUpper === void 0 ? void 0 : onScrollToUpper();
                    } })),
            }) })) : isNeedPrevLoadMore ? (_jsx(Spinning, { text: i18n.t('messageListLoading'), className: styles.loading, size: 'small' })) : null }));
};
export const ChatScrollView = ({ children }) => {
    const refScroll = useRef(null);
    const { upperThreshold, onScrollToUpper } = useScrollMore();
    const { isUnshiftingMessageFlag, clearUnshiftingMessageFlg } = useConversationStore(store => ({
        isUnshiftingMessageFlag: store.isUnshiftingMessageFlag,
        clearUnshiftingMessageFlg: store.clearUnshiftingMessageFlg,
    }));
    const hasProcessChatMessageGroup = useConversationStore(store => !!store.inProcessChatMessageGroup);
    const chatEvent = useUiEventStore(store => store.event);
    useEffect(() => {
        if (isUnshiftingMessageFlag) {
            clearUnshiftingMessageFlg();
        }
    }, [isUnshiftingMessageFlag]);
    const chatContent = useMemo(() => (_jsxs(_Fragment, { children: [_jsx(ChatTopStatus, Object.assign({}, { onScrollToUpper })), children] })), [children, onScrollToUpper]);
    useEffect(() => {
        chatEvent.on(UIEventType.ChatSlotToBottom, () => {
            var _a;
            (_a = refScroll.current) === null || _a === void 0 ? void 0 : _a.scrollToBottom();
        });
        chatEvent.on(UIEventType.ChatSlotScrollToAnchorBottom, () => {
            var _a;
            (_a = refScroll.current) === null || _a === void 0 ? void 0 : _a.scrollToAnchorBottom();
        });
        chatEvent.on(UIEventType.ChatSlotRemoveAnchorBottom, () => {
            var _a;
            (_a = refScroll.current) === null || _a === void 0 ? void 0 : _a.removeAnchorBottom();
        });
    }, []);
    useUpdateEffect(() => {
        var _a, _b;
        if (hasProcessChatMessageGroup) {
            (_a = refScroll.current) === null || _a === void 0 ? void 0 : _a.scrollToAnchorBottom();
        }
        else {
            (_b = refScroll.current) === null || _b === void 0 ? void 0 : _b.removeAnchorBottom();
        }
    }, [hasProcessChatMessageGroup]);
    logger.debug('ScrollView props isUnshiftingMessageFlag', {
        isUnshiftingMessageFlag,
    });
    return (_jsx(ScrollView, Object.assign({ ref: refScroll, className: cls(styles['scroll-container']), scrollY: true, isLoadMore: isUnshiftingMessageFlag, lowerThreshold: upperThreshold, onScrollToLower: onScrollToUpper, isNeedHelper: true }, { children: chatContent })));
};
