// @coze/chat-sdk/dist/lib-source/src/libs/ui-kit/atomic/md-stream/context/index.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useMemo, } from 'react';
import { usePersistCallback } from "../../../../hooks";
const MdStreamPropsContext = createContext({});
export const ChatFamePropsProvider = ({ markdown, children, onImageClick, i18n, logger, selectable, taskDisabled, enableHtmlTags, enableCodeBy4Space, onTaskChangeHandle, eventCallbacks, imgClassName, imgStyle, }) => {
    const getMarkdown = usePersistCallback(() => markdown || '');
    const value = useMemo(() => ({
        onImageClick,
        i18n,
        logger,
        selectable,
        taskDisabled,
        enableHtmlTags,
        enableCodeBy4Space,
        onTaskChangeHandle,
        getMarkdown,
        eventCallbacks,
        imgClassName,
        imgStyle,
    }), [
        onImageClick,
        i18n,
        logger,
        selectable,
        taskDisabled,
        enableHtmlTags,
        enableCodeBy4Space,
        onTaskChangeHandle,
        getMarkdown,
        eventCallbacks,
        imgClassName,
        imgStyle,
    ]);
    return (_jsx(MdStreamPropsContext.Provider, Object.assign({ value: value }, { children: children })));
};
export const useMdStreamContext = () => useContext(MdStreamPropsContext);
export const useMdStreamI18n = () => {
    const context = useMdStreamContext();
    const { i18n } = context || {};
    return i18n;
};
export const useMdStreamLogger = () => {
    const context = useMdStreamContext();
    const { logger } = context || {};
    return logger;
};
export const useMdStreamSelectable = () => {
    const context = useMdStreamContext();
    const { selectable } = context || {};
    return selectable;
};
export const useMdStreamTaskDisabled = () => {
    const context = useMdStreamContext();
    const { taskDisabled } = context || {};
    return taskDisabled;
};
export const useMdStreamTaskChangeHandle = () => {
    const context = useMdStreamContext();
    const { onTaskChangeHandle } = context || {};
    return onTaskChangeHandle;
};
export const useMdStreamEnableHtmlTags = () => {
    const context = useMdStreamContext();
    const { enableHtmlTags } = context || {};
    return enableHtmlTags;
};
export const useMdStreamEnableCodeBy4Space = () => {
    const context = useMdStreamContext();
    const { enableCodeBy4Space } = context || {};
    return enableCodeBy4Space;
};
export const useMdStreamGetMarkdown = () => {
    const context = useMdStreamContext();
    const { getMarkdown } = context || {};
    return (getMarkdown === null || getMarkdown === void 0 ? void 0 : getMarkdown()) || '';
};
