// @coze/chat-sdk/dist/lib-source/src/libs/provider/context/chat-frame-props-context.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext } from 'react';
import { useValidContext } from "../../hooks";
const ChatFramePropsContext = createContext({
    chat: null,
    auth: null,
    user: null,
    children: null,
    setting: null,
});
export const ChatFamePropsProvider = (_a) => {
    var { children } = _a, props = __rest(_a, ["children"]);
    return _jsx(ChatFramePropsContext.Provider, { value: props, children: children });
};
export const useChatPropsContext = () => useValidContext(ChatFramePropsContext);
