// @coze/chat-sdk/dist/lib-source/src/libs/framework/component/chat-group-list/prologue-group/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useMemo } from 'react';
import { View, Text } from '@tarojs/components';
import { Spacing } from "../../../../ui-kit";
import { useCreatePrologueMessage } from "../../../../services";
import { useConversationStore, useI18n } from "../../../../provider";
import { ChatGroup } from '../chat-group';
import styles from './index.module.less';
// section分隔线以及
export const SectionPrologueGroup = ({ chatMessageGroupNow, chatMessageGroupNext }) => {
    const { hasInProcessChatMessageGroup, sectionId } = useConversationStore(store => ({
        hasInProcessChatMessageGroup: !!store.inProcessChatMessageGroup,
        sectionId: store.sectionId,
    }));
    const isShowDivide = useMemo(() => {
        if (chatMessageGroupNext) {
            if (chatMessageGroupNext.sectionId !== chatMessageGroupNow.sectionId) {
                return true;
            }
        }
        else if (chatMessageGroupNow.sectionId !== sectionId) {
            return true;
        }
        return false;
    }, [chatMessageGroupNow, chatMessageGroupNext, sectionId]);
    const isShowSuggestion = useMemo(() => {
        if (isShowDivide &&
            !hasInProcessChatMessageGroup &&
            !chatMessageGroupNext) {
            return true;
        }
        return false;
    }, [
        isShowDivide,
        hasInProcessChatMessageGroup,
        chatMessageGroupNext,
        sectionId,
    ]);
    const nextSectionId = (chatMessageGroupNext === null || chatMessageGroupNext === void 0 ? void 0 : chatMessageGroupNext.sectionId) || sectionId;
    if (!isShowDivide) {
        return null;
    }
    return (_jsx(PrologueChatMessage, { isShowDivide: isShowDivide, isShowSuggestion: isShowSuggestion, nextSectionId: nextSectionId }));
};
const PrologueChatMessage = ({ isShowDivide, isShowSuggestion, nextSectionId }) => {
    const prologueChatGroup = useCreatePrologueMessage(nextSectionId);
    const i18n = useI18n();
    return (_jsxs(View, { children: [isShowDivide ? (_jsxs(Spacing, Object.assign({ gap: 20, className: styles['divide-container'], verticalCenter: true, horizontalCenter: true }, { children: [_jsx(View, { className: styles['divide-line'] }), _jsx(Text, { children: i18n.t('clearContextDivide') }), _jsx(View, { className: styles['divide-line'] })] }))) : null, prologueChatGroup ? (_jsx(ChatGroup, { chatGroup: prologueChatGroup, isShowSuggestion: isShowSuggestion })) : null] }));
};
export const TopPrologueGroup = () => {
    const { isShowOnBoarding, prevHasMore } = useConversationStore(store => ({
        isShowOnBoarding: store.isShowOnBoarding,
        chatMessageGroups: store.chatMessageGroups,
        inProcessChatMessageGroup: store.inProcessChatMessageGroup,
        prevHasMore: store.prevHasMore,
    }));
    if (isShowOnBoarding() || prevHasMore) {
        return null;
    }
    return _jsx(PrologueChatMessage, { isShowDivide: false, isShowSuggestion: false });
};
