// @coze/chat-sdk/dist/lib-source/src/libs/provider/store/chat-status.js
import { useMemo } from 'react';
import { create } from 'zustand';
import { useUpdateEffect } from "../../hooks";
import { useChatPropsContext } from '../context';
const createChatStatusStore = ({ isReadonly }) => create()((set, get) => {
    const wrapStateFunc = (func) => (...args) => {
        const result = func(...args);
        get().setOpDisabledState();
        return result;
    };
    return {
        isReadonly: isReadonly || false,
        isDeleting: false,
        isSendingMsg: false,
        isClearingContext: false,
        isAudioRecording: false,
        disableState: {
            clearMessage: isReadonly || false,
            clearContext: isReadonly || false,
            input: isReadonly || false,
        },
        setIsReadonly: wrapStateFunc((isReadonlyNew) => {
            set({
                isReadonly: isReadonlyNew,
            });
        }),
        setIsAudioRecording: wrapStateFunc((isAudioRecording) => {
            set({
                isAudioRecording,
            });
        }),
        setIsDeleting: wrapStateFunc((isDeleting) => {
            set({
                isDeleting,
            });
        }),
        setIsSendingMsg: wrapStateFunc((isSendingMsg) => {
            set({
                isSendingMsg,
            });
        }),
        setIsClearingContext: wrapStateFunc((isClearingContext) => {
            set({
                isClearingContext,
            });
        }),
        setOpDisabledState: () => {
            set({
                disableState: get().getOpDisabledState(),
            });
        },
        getOpDisabledState: () => {
            const { isDeleting, isSendingMsg, isClearingContext, isReadonly: isReadonlyNew, } = get();
            const isInteracting = isDeleting || isSendingMsg || isClearingContext;
            return {
                clearMessage: isReadonlyNew || isInteracting,
                input: isReadonlyNew || isInteracting,
                clearContext: isReadonlyNew || isInteracting,
            };
        },
    };
});
export const useCreateStatusStore = () => {
    var _a;
    const chatProps = useChatPropsContext();
    const chatStatusStore = useMemo(() => { var _a; return createChatStatusStore({ isReadonly: (_a = chatProps.ui) === null || _a === void 0 ? void 0 : _a.isReadonly }); }, []);
    useUpdateEffect(() => {
        var _a;
        chatStatusStore.getState().setIsReadonly(!!((_a = chatProps.ui) === null || _a === void 0 ? void 0 : _a.isReadonly));
    }, [(_a = chatProps.ui) === null || _a === void 0 ? void 0 : _a.isReadonly]);
    return chatStatusStore;
};
