// @coze/chat-sdk/dist/lib-source/src/libs/provider/store/chat-info.js
import { useMemo } from 'react';
import { create } from 'zustand';
import { logger } from "../../utils";
import { useUpdateEffect } from "../../hooks";
import { useChatPropsContext } from '../context';
const getCustomChatInfo = (chat) => Object.fromEntries(Object.entries({
    name: chat.name,
    description: chat.description,
    icon_url: chat.icon_url,
    onboarding_info: chat.onboarding_info,
    bgInfo: chat.bgInfo,
    voiceInfo: chat.voiceInfo,
    suggestPromoteInfo: chat.suggestPromoteInfo,
}).filter(([_, value]) => !!value));
const combineChatInfo = (chatInfoFromInit, customChatInfo) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    const customSuggestionList = ((_b = (_a = customChatInfo.onboarding_info) === null || _a === void 0 ? void 0 : _a.suggested_questions) === null || _b === void 0 ? void 0 : _b.filter(item => !!item)) || [];
    return Object.assign(Object.assign(Object.assign({}, chatInfoFromInit), customChatInfo), { onboarding_info: {
            display_all_suggestions: (_d = (_c = customChatInfo.onboarding_info) === null || _c === void 0 ? void 0 : _c.display_all_suggestions) !== null && _d !== void 0 ? _d : (_e = chatInfoFromInit.onboarding_info) === null || _e === void 0 ? void 0 : _e.display_all_suggestions,
            prologue: ((_f = customChatInfo.onboarding_info) === null || _f === void 0 ? void 0 : _f.prologue)
                ? (_g = customChatInfo.onboarding_info) === null || _g === void 0 ? void 0 : _g.prologue
                : (_h = chatInfoFromInit.onboarding_info) === null || _h === void 0 ? void 0 : _h.prologue,
            suggested_questions: (_k = ((customSuggestionList === null || customSuggestionList === void 0 ? void 0 : customSuggestionList.length)
                ? customSuggestionList
                : (_j = chatInfoFromInit.onboarding_info) === null || _j === void 0 ? void 0 : _j.suggested_questions)) === null || _k === void 0 ? void 0 : _k.filter(item => !!item),
        } });
};
const createChatInfoStore = ({ chat }) => {
    // If the chat's parameter(name, description, icon_url, onboarding_info) is not empty, then use it to set the chat info
    let customChatInfo = getCustomChatInfo(chat);
    let chatInfoFromInit;
    return create()(set => ({
        id: chat.appId,
        type: chat.type,
        info: null,
        isLoading: true,
        error: null,
        setIsLoading: (isLoading) => {
            set({
                isLoading,
                error: null,
            });
        },
        // 设置系统配置的数据
        setChatInfo: (info) => {
            chatInfoFromInit = info;
            logger.debug('setChatInfo', { initInfo: info, customChatInfo });
            set({
                error: null,
                isLoading: false,
                info: combineChatInfo(chatInfoFromInit, customChatInfo),
            });
        },
        // 设置自定义参数
        setCustomChatInfo: (info) => {
            customChatInfo = getCustomChatInfo(info);
            if (!chatInfoFromInit) {
                return;
            }
            logger.debug('setCustomChatInfo', {
                oldInfo: chatInfoFromInit,
                customChatInfo,
            });
            set({
                info: combineChatInfo(chatInfoFromInit, customChatInfo),
            });
        },
        setError: (error) => {
            set({
                isLoading: false,
                error,
            });
        },
    }));
};
export const useCreateChatInfoStore = () => {
    const chatProps = useChatPropsContext();
    const chatInfoStore = useMemo(() => createChatInfoStore({ chat: chatProps.chat }), []);
    useUpdateEffect(() => {
        logger.debug('useCreateChatInfoStore update chat props: ', chatProps.chat);
        chatInfoStore.getState().setCustomChatInfo(chatProps.chat);
    }, [chatProps.chat]);
    return chatInfoStore;
};
