// @coze/chat-sdk/dist/lib-source/src/libs/ui-kit/atomic/upload/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useRef } from 'react';
import Taro from '@tarojs/taro';
import { View } from '@tarojs/components';
import { getFileTypeByFile, isWeb, logger, showToast } from "../../../utils";
import { FileTypeEnum } from "../../../types";
import styles from './index.module.less';
export const Upload = ({ children, accept, onChooseFile, frameEventTarget }) => {
    const ref = useRef(null);
    const onWebInputFileChange = (e) => {
        const { files } = e.target;
        if (onChooseFile) {
            if (files && files.length > 0) {
                const fileInfos = [];
                for (const file of files) {
                    const fileType = getFileTypeByFile(file);
                    fileInfos.push({
                        from: 'H5_Input_Chooser',
                        type: fileType,
                        size: file.size,
                        file,
                        tempFilePath: URL.createObjectURL(file),
                    });
                }
                onChooseFile === null || onChooseFile === void 0 ? void 0 : onChooseFile(fileInfos);
            }
            else {
                // 没有选中文件
            }
        }
        e.target.value = '';
    };
    return (_jsxs(View, Object.assign({ className: styles.container }, { children: [isWeb && (_jsx("input", { type: "file", multiple: false, accept: accept, autoComplete: "off", tabIndex: -1, className: styles.input, onChange: onWebInputFileChange, ref: ref })), _jsx(View, Object.assign({ onClick: () => {
                    var _a, _b;
                    if (isWeb) {
                        if (ref.current) {
                            (_b = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.click) === null || _b === void 0 ? void 0 : _b.call(_a);
                        }
                    }
                    else {
                        try {
                            Taro.chooseMedia({
                                count: 1,
                                mediaType: ['image'],
                                sizeType: ['original', 'compressed'],
                                sourceType: ['album', 'camera'],
                                success(res) {
                                    logger.debug('chooseMedia success:', res);
                                    const fileInfos = res.tempFiles.map(item => ({
                                        from: 'Taro_Image_Chooser',
                                        type: 
                                        // @ts-expect-error -- linter-disable-autofix
                                        item.fileType === 'image' || item.mediaType === 'image'
                                            ? FileTypeEnum.IMAGE
                                            : FileTypeEnum.VIDEO,
                                        size: item.size,
                                        tempFilePath: item.tempFilePath,
                                        file: {
                                            filePath: item.tempFilePath,
                                        },
                                    }));
                                    onChooseFile === null || onChooseFile === void 0 ? void 0 : onChooseFile(fileInfos);
                                },
                                fail(res) {
                                    logger.error('chooseMedia fail:', res);
                                    if (res.errMsg.includes('cancel')) {
                                        return;
                                    }
                                    else {
                                        showToast({
                                            content: `选择文件失败: ${res.errMsg}`,
                                            icon: 'error',
                                        }, frameEventTarget);
                                    }
                                },
                            });
                        }
                        catch (error) {
                            logger.error('chooseMedia error:', error);
                        }
                    }
                } }, { children: children }))] })));
};
