// @coze/chat-sdk/dist/lib-source/src/libs/framework/component/chat-background/helper.js
export const cropperSizeMap = {
    pc: {
        width: 486,
        height: 346,
    },
    mobile: {
        width: 248,
        height: 346,
    },
};
function hexToRgb(hex) {
    // 如果是三位十六进制，将其扩展为六位
    if (hex.length === 4) {
        hex = hex[1] + hex[1] + hex[2] + hex[2] + hex[3] + hex[3];
    }
    hex = hex.slice(1); // 去掉'#'
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4), 16);
    return `rgb(${r},${g},${b})`;
}
// 输入透明度系数 和color 返回新的颜色
export function addAlpha(color, alpha) {
    if (/^((#[0-9A-Fa-f]{3})|(#[0-9A-Fa-f]{6}))$/.test(color)) {
        color = hexToRgb(color);
    }
    const regex = /^rgba\((\d{1,3}),(\d{1,3}),(\d{1,3})\)$/;
    if (!regex.test(color)) {
        return color;
    }
    const values = color.slice(5, -1).split(',');
    values.push(alpha.toString());
    const newColor = `rgba(${values.join(',')})`;
    return newColor;
}
export const getGradient = (gradient, canvasData, cropperWidth = 0) => {
    const { left: cropperImgLeft = 0, width: cropperImgWidth = 0 } = canvasData;
    if (cropperImgWidth && cropperWidth) {
        return {
            left: cropperImgLeft / cropperWidth,
            right: (cropperWidth - cropperImgWidth - cropperImgLeft) / cropperWidth,
        };
    }
    else {
        return gradient;
    }
};
export const getDeviceMode = (width, height) => {
    if (!width || !height) {
        return 'pc';
    }
    return width / height <= getStandardRatio('mobile') ? 'mobile' : 'pc';
};
export const getStandardRatio = (mode) => cropperSizeMap[mode].width / cropperSizeMap[mode].height;
