// @coze/chat-sdk/dist/lib-source/src/libs/provider/hooks/use-voice-info.js
import { logger } from "../../utils";
import { InputType } from "../../ui-kit/chat-input/type";
import { Language } from "../../types";
import { useChatInfoStore, useChatPropsStore, useI18n } from '../context';
export const useVoiceInfo = () => {
    const voiceInfo = useChatInfoStore(store => { var _a; return (_a = store.info) === null || _a === void 0 ? void 0 : _a.voiceInfo; });
    return voiceInfo;
};
export const useIsNeedAudioInput = () => {
    const isNeedAudio = useChatPropsStore(store => { var _a, _b, _c; return (_c = (_b = (_a = store.ui) === null || _a === void 0 ? void 0 : _a.chatSlot) === null || _b === void 0 ? void 0 : _b.input) === null || _c === void 0 ? void 0 : _c.isNeedAudio; });
    /*const isWeixinWeb = useMemo(() => {
      if (isWeb) {
        if (navigator.userAgent.includes("wechat")) {
          return true;
        }
      }
      return false;
    }, []);*/
    // Tha audio input is not suppotted in weixin webview
    return isNeedAudio || false;
};
export const useDefaultInputMode = () => {
    const defaultMode = useChatInfoStore(store => { var _a, _b; return (_b = (_a = store.info) === null || _a === void 0 ? void 0 : _a.inputMode) === null || _b === void 0 ? void 0 : _b.default; });
    const isNeedAudio = useIsNeedAudioInput();
    return defaultMode === 2 && isNeedAudio ? InputType.Voice : InputType.Text;
};
export const useVoiceId = () => {
    var _a, _b, _c, _d;
    const voiceInfo = useChatInfoStore(store => { var _a; return (_a = store.info) === null || _a === void 0 ? void 0 : _a.voiceInfo; });
    const i18n = useI18n();
    const langList = i18n.language === Language.ZH_CN ? ['zh', 'en'] : ['en', 'zh'];
    let voiceId;
    for (let i = 0; i < langList.length; i++) {
        if ((_b = (_a = voiceInfo === null || voiceInfo === void 0 ? void 0 : voiceInfo.voiceConfigMap) === null || _a === void 0 ? void 0 : _a[langList[i]]) === null || _b === void 0 ? void 0 : _b.voice_id) {
            voiceId = (_d = (_c = voiceInfo === null || voiceInfo === void 0 ? void 0 : voiceInfo.voiceConfigMap) === null || _c === void 0 ? void 0 : _c[langList[i]]) === null || _d === void 0 ? void 0 : _d.voice_id;
        }
    }
    logger.debug('useVoiceId voiceInfo:', voiceInfo, voiceId);
    return voiceId;
};
export const useIsNeedTextToAudio = () => {
    const voiceInfo = useVoiceInfo();
    const voiceId = useVoiceId();
    logger.debug('useIsNeedAudioInput info:', {
        voiceId,
        isNeedTextToAudio: (voiceId && !!(voiceInfo === null || voiceInfo === void 0 ? void 0 : voiceInfo.isTextToVoiceEnable)) || false,
    });
    return (voiceId && !!(voiceInfo === null || voiceInfo === void 0 ? void 0 : voiceInfo.isTextToVoiceEnable)) || false;
};
