// @coze/chat-sdk/dist/lib-source/src/libs/utils/mini-chat-error.js
export class MiniChatError extends Error {
    constructor(code, msg) {
        super(msg);
        this.code = code;
        this.msg = msg;
    }
    getErrorMessageByI18n(i18n, options, fallbackText) {
        return i18n.t(`miniChatError_${this.code}`, options, fallbackText || this.msg);
    }
}
export const convertToMinChatError = (error) => {
    const miniChatError = error;
    return new MiniChatError((miniChatError === null || miniChatError === void 0 ? void 0 : miniChatError.code) || -1, (miniChatError === null || miniChatError === void 0 ? void 0 : miniChatError.msg) ||
        (miniChatError === null || miniChatError === void 0 ? void 0 : miniChatError.message) ||
        (miniChatError === null || miniChatError === void 0 ? void 0 : miniChatError.stack) ||
        'unknown error');
};
export var MiniChatErrorCode;
(function (MiniChatErrorCode) {
    MiniChatErrorCode[MiniChatErrorCode["Unknown"] = -1] = "Unknown";
    MiniChatErrorCode[MiniChatErrorCode["Audio_Permission_Denied"] = 1000] = "Audio_Permission_Denied";
    MiniChatErrorCode[MiniChatErrorCode["Audio_Translation_NoContent"] = 1001] = "Audio_Translation_NoContent";
    MiniChatErrorCode[MiniChatErrorCode["Custom_Conversation_Create_Error"] = 1002] = "Custom_Conversation_Create_Error";
    MiniChatErrorCode[MiniChatErrorCode["Custom_GetMessageList_Error"] = 10003] = "Custom_GetMessageList_Error";
    MiniChatErrorCode[MiniChatErrorCode["SDK_API_OVERDUE_PAYMENT"] = 4011] = "SDK_API_OVERDUE_PAYMENT";
    MiniChatErrorCode[MiniChatErrorCode["SDK_API_APP_UnPublished"] = 4015] = "SDK_API_APP_UnPublished";
})(MiniChatErrorCode || (MiniChatErrorCode = {}));
