// @coze/chat-sdk/dist/lib-source/src/libs/framework/component/chat-group-list/command/command-copy/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import cls from 'classnames';
import Taro from '@tarojs/taro';
import { Text } from '@tarojs/components';
import { setClipboardData, showToast } from "../../../../../utils";
import { SvgCopy } from "../../../../../ui-kit/atomic/svg";
import { DisableContainer } from "../../../../../ui-kit/atomic/disable-container";
import { IconButton, Spacing } from "../../../../../ui-kit";
import { UIEventType } from "../../../../../types";
import { useI18n, useUiEventStore, } from "../../../../../provider/context/chat-store-context";
import styles from './index.module.less';
export const CommandCopy = ({ text, isShowText, svgTheme, isNeedHover, className }) => {
    const targetEventCenter = useUiEventStore(store => store.event);
    const i18n = useI18n();
    return (_jsx(DisableContainer, Object.assign({ className: cls(styles.container, className) }, { children: _jsxs(Spacing, Object.assign({ width100: true, onClick: event => {
                event.stopPropagation();
                Taro.eventCenter.trigger(UIEventType.FrameClick);
                setClipboardData({
                    data: text,
                    success(isUseWeb) {
                        if (isUseWeb) {
                            showToast({
                                content: i18n.t('copySuccess'),
                                icon: 'success',
                            }, targetEventCenter);
                        }
                    },
                    fail() {
                        showToast({
                            content: i18n.t('copyFailed'),
                            icon: 'error',
                        }, targetEventCenter);
                    },
                });
            }, gap: 8 }, { children: [_jsx(IconButton, Object.assign({ size: "small", type: "square-hover-btn", hoverTheme: isNeedHover ? 'hover' : 'none' }, { children: _jsx(SvgCopy, { theme: svgTheme }) })), isShowText ? (_jsx(Text, Object.assign({ className: styles.text, numberOfLines: 1, maxLines: 1, overflow: "ellipsis" }, { children: i18n.t('copyText') }))) : null] })) })));
};
