// @coze/chat-sdk/dist/lib-source/src/libs/utils/recorder-manager/taro-recorder-manager.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import Taro from '@tarojs/taro';
import { BaseRecorderManager, RecorderEvent, } from './type';
import { MiniChatError, MiniChatErrorCode } from '../mini-chat-error';
import { logger } from '../logger';
import { isTT } from '../device';
export class TaroRecorderManager extends BaseRecorderManager {
    constructor() {
        super(...arguments);
        this.options = {};
    }
    start(options) {
        var _a, _b, _c, _d, _e, _f;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.isStop) {
                throw new MiniChatError(-1, 'recorder is stop');
            }
            logger.debug('TaroRedorManager start instance stopFlag', (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.isStop);
            // There's only one instance at a time.
            if (!((_b = TaroRecorderManager.curInstance) === null || _b === void 0 ? void 0 : _b.isStop)) {
                (_c = TaroRecorderManager.curInstance) === null || _c === void 0 ? void 0 : _c.stop();
            }
            TaroRecorderManager.curInstance = this;
            logger.debug('TaroRedorManager start instance recorder', TaroRecorderManager.recorder);
            if (!TaroRecorderManager.recorder) {
                TaroRecorderManager.recorder = Taro.getRecorderManager();
                this.registerEvent();
            }
            this.format = isTT ? 'wav' : 'mp3';
            this.fileName = `recorder_${Date.now()}.${this.format}`;
            this.options = options;
            const { sampleRate = 8000, numberOfChannels = 1 } = this.options;
            try {
                yield this.checkPermission();
                if (this.isStop) {
                    (_d = TaroRecorderManager.curInstance) === null || _d === void 0 ? void 0 : _d.emit(RecorderEvent.STOP, {
                        duration: 0,
                        tempFilePath: '',
                        fileSize: 0,
                    });
                    (_e = TaroRecorderManager.recorder) === null || _e === void 0 ? void 0 : _e.stop();
                    return;
                }
                TaroRecorderManager.recorder.start({
                    numberOfChannels,
                    sampleRate,
                    format: this.format,
                    frameSize: 2,
                });
                const getVolume = (_f = TaroRecorderManager.curInstance) === null || _f === void 0 ? void 0 : _f.genGetVolumeFunc();
                getVolume === null || getVolume === void 0 ? void 0 : getVolume();
            }
            catch (err) {
                logger.error('TaroRecorderManager start error', err);
                this.emitError(new MiniChatError(MiniChatErrorCode.Audio_Permission_Denied, 'Permission Denied'));
            }
        });
    }
    checkPermission() {
        return new Promise((resolve, reject) => {
            Taro.getSetting({
                success: res => {
                    logger.debug('TaroRecorderManager getSetting', res);
                    if (res.authSetting['scope.record'] !== true) {
                        if (isTT) {
                            resolve(true);
                        }
                        else {
                            Taro.authorize({
                                scope: 'scope.record',
                                success: () => {
                                    logger.debug('TaroRecorderManager authorize scope.record true');
                                    resolve(true);
                                },
                                fail: resFail => {
                                    logger.error('TaroRecorderManager authorize scope.record false', resFail);
                                    reject(new Error('TaroRecorderManager authorize scope.record false'));
                                },
                            });
                        }
                    }
                    else {
                        logger.debug('TaroRecorderManager getSetting scope.record true');
                        resolve(true);
                    }
                },
                fail: () => {
                    logger.debug('TaroRecorderManager getSetting fails');
                    resolve(true);
                },
            });
        });
    }
    registerEvent() {
        var _a, _b, _c, _d, _e, _f, _g;
        (_a = TaroRecorderManager.recorder) === null || _a === void 0 ? void 0 : _a.onStart(() => {
            var _a;
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emit(RecorderEvent.START, {});
        });
        (_b = TaroRecorderManager.recorder) === null || _b === void 0 ? void 0 : _b.onPause(() => {
            var _a;
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emit(RecorderEvent.PAUSE, {});
        });
        (_c = TaroRecorderManager.recorder) === null || _c === void 0 ? void 0 : _c.onResume(() => {
            var _a;
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emit(RecorderEvent.RESUME, {});
        });
        (_d = TaroRecorderManager.recorder) === null || _d === void 0 ? void 0 : _d.onStop(event => {
            var _a;
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emit(RecorderEvent.STOP, {
                duration: event.duration,
                tempFilePath: event.tempFilePath,
                fileSize: event.fileSize,
                fileName: this.fileName,
            });
        });
        (_f = (_e = TaroRecorderManager.recorder) === null || _e === void 0 ? void 0 : _e.onInterruptionBegin) === null || _f === void 0 ? void 0 : _f.call(_e, () => {
            var _a;
            logger.debug('TaroRecorderManager onInterruptionBegin');
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emit(RecorderEvent.INTERRUPT, {});
        });
        (_g = TaroRecorderManager.recorder) === null || _g === void 0 ? void 0 : _g.onError(event => {
            var _a, _b;
            logger.error('TaroRecorderManager onError', event);
            (_a = TaroRecorderManager.curInstance) === null || _a === void 0 ? void 0 : _a.emitError(new MiniChatError(((_b = event === null || event === void 0 ? void 0 : event.errMsg) === null || _b === void 0 ? void 0 : _b.includes('deny'))
                ? MiniChatErrorCode.Audio_Permission_Denied
                : -1, event.errMsg));
        });
    }
    genGetVolumeFunc() {
        const getVolume = () => {
            if (this.isStop) {
                this.emit(RecorderEvent.VOLUME, {
                    volume: 0,
                });
                return;
            }
            const volume = Math.random() * 0.5;
            this.emit(RecorderEvent.VOLUME, {
                volume: volume < 0.002 ? 0 : volume,
            });
            setTimeout(() => {
                getVolume();
            }, 100);
        };
        return getVolume;
    }
    emitError(error) {
        this.emit(RecorderEvent.ERROR, error);
    }
    stop() {
        var _a;
        if (this.isStop) {
            logger.warn('recorder has been stopped');
            return;
        }
        this.isStop = true;
        try {
            (_a = TaroRecorderManager.recorder) === null || _a === void 0 ? void 0 : _a.stop();
        }
        catch (err) {
            this.emitError(new MiniChatError(-1, 'unknown error'));
        }
    }
    pause() {
        var _a;
        if (this.isStop) {
            logger.warn('recorder has been stopped');
            return;
        }
        (_a = TaroRecorderManager.recorder) === null || _a === void 0 ? void 0 : _a.pause();
    }
    resume() {
        var _a;
        if (this.isStop) {
            logger.warn('recorder has been stopped');
            return;
        }
        (_a = TaroRecorderManager.recorder) === null || _a === void 0 ? void 0 : _a.resume();
    }
    destroy() {
        this.stop();
        this.event.off();
    }
}
TaroRecorderManager.recorder = null;
